DAC REV-B Firmware

BrandtBuchda opened the pull request:
This PR is to merge the DAC REV-B board's IP and firmware into develop as well as the inverter status mux IP. This PR contains the necessary IP block and user application firmware to drive the [DAC expansion board](https://github.com/Severson-Group/AMDC-Hardware/pull/133). It contains a user app that generates a 3-phase power waveform on channels 0-2 and implements commands that can be used to configure and drive the DAC in real time. It also contains the IP block that maps the status lines on the inverter ports to any arbitrary device driver in the FPGA and the C driver that configures the mux on initialization.

The changes made include changes to the amdc_revd block diagram. The inverter status mux and DAC IP blocks were added to the block diagram and connected to the inverter. 

A docs folder was added at [/sdk/bare/common/drv/docs](https://github.com/Severson-Group/AMDC-Firmware/tree/dac-revb-firmware/sdk/bare/common/drv/docs)  with markdown documentation of the C drivers used to drive the DAC and status mux.

The documentation for the DAC ip_block can be found [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/ip_repo/amdc_dac_2.0/README.md).

The documentation for the firmware can be found [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/usr/dac/README.md).

The documentation for the inverter status mux ip_block can be found [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/ip_repo/amdc_inv_status_mux_1.0/README.md).

The documentation for the inverter status mux firmware can be found [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/drv/docs/Status-Mux.md).

BrandtBuchda committed:
Added DAC-REVB IP block and firmware

BrandtBuchda committed:
Added external trigger to IP block

BrandtBuchda committed:
Reverted changes to amdc_revd.bd and user_config.h

BrandtBuchda committed:
Updated Readme

npetersen2 reviewed (changes requested):
@BrandtBuchda Nice work, a few comments / requests:

1. You define `dac_reg_t` enum -- can you prefix the enum values with `DAC_REG_***`. This will help the DAC code not conflict with other code in the system.
2. The link to `task_dac.c` in the app readme file doesn't work.
3. In your `task_dac.c`, can you make the `FREQ` variable static so it isn't global?
4. I believe a `MAX()` macro like you have defined is available in `sys/util.h`, please don't redefine it here.
5. I don't think `task_dac.c` needs the `drv/led.h` include, please remove.
6. In your `APP_DAC` README file, can you add a section that explains how to use the commands from a user perspective? This should be a *small* summary from the DAC datasheet about how it works, the different options a user has to interact with the DAC, etc. This section should be written such that when someone starts using the DAC, they read it and know all the knowledge for how to use the DAC.. i.e. can they just call `dac_set_voltage(ch, v)` and it will behave correct? Is there any major configuration they should know about? Think about this from @mimayberry perspective -- he will start using your DAC platform soon and will need to know how to correctly use it; your doc file should be enough that he never has to ask you questions!

npetersen2 commented:
Lastly, we require all merges into `develop` to pass the auto formatting checks (you can see a red X for this PR which indicates the check failed). Please format the code and make sure the red X turns into a green check mark.

You can see what failed in the check by looking at this: https://github.com/Severson-Group/AMDC-Firmware/pull/153/checks?check_run_id=1068864213

You can read about the auto formatting tool here: https://github.com/Severson-Group/AMDC-Firmware/blob/develop/CONTRIBUTING.md#code-formatting

BrandtBuchda committed:
Formatted files

BrandtBuchda committed:
Fixed bsp formattting

BrandtBuchda committed:
Added config information to README

BrandtBuchda committed:
Added Status Mux IP

BrandtBuchda committed:
Changed device numbering (1-8)

BrandtBuchda committed:
Formatted files

BrandtBuchda committed:
Reverted constraints_amdc_revd.xdc

npetersen2 commented:
@BrandtBuchda Let's try to get this merged into `develop` ASAP. I see that the block diagram has been updated since you branched from `develop` -- unfortunately, there is no merge tool for the block diagram in Vivado. You must checkout out the latest `develop` block diagram and remake your required changes (i.e. re-add your IP blocks). Note that the IP memory map addresses may change, so update your C drivers accordingly.

npetersen2 commented:
@BrandtBuchda , we can also consider merging just the FPGA IP + C driver into `develop`.

Then, we can open an issue about the function generator application for you complete soon.

BrandtBuchda committed:
Added sts mux IP readme

BrandtBuchda committed:
merge changes from develop

BrandtBuchda committed:
Added inv status mux and dac IP to block diagram

BrandtBuchda committed:
Formatted drivers

BrandtBuchda committed:
Fixed sts_mux.h formatting

BrandtBuchda committed:
Reverted changes to amdc constraints

BrandtBuchda committed:
Added driver doc

BrandtBuchda committed:
Fixed links in docs

BrandtBuchda committed:
Added DAC driver doc

BrandtBuchda committed:
Added datasheet links

npetersen2 reviewed (changes requested):
@BrandtBuchda I have a few comments on the docs, overall looking good.

1. In [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/ip_repo/amdc_dac_2.0/README.md), you first state `DAC_CONFIG1` is unused, then address it later on. I think this is a mistake. Also, can you add a section for the other registers after the main table and just state that the IP is designed such that these mimic the device exactly.
2. In [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/usr/dac/README.md), you talk about the example 3-phase voltage output on channels 0,2,4 and somewhere else I thought i saw 0,1,2. Can this just be configurable from the user? If not, please make the docs consistent.
3. In [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/usr/dac/README.md), can you add a subheader between the commands related to the 3-phase output and lower-level commands? I think it is confusing to list them all as one block of commands. Consider creating a whole new command (e.g. "dacex") for the example 3-ph voltages (this new command idea might not be necessary after reading the whole doc file).
4. You mention that you must set the `GAIN` register to the correct `0x01FF`. What if the user doesn't do this? Does it break things? Are there other things that you should draw the user's attention to like this?
5. Please go back through the docs and update all places that you use numbers, register names, etc to use the code formatting in markdown, i.e. `this is code` => "example `GAIN` register is `0xFFF`" Also, be sure to use the code styling markdown (as seen below) to make the docs pretty. i.e. "```C"
6. In [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/usr/dac/task_dac.c#L87), I'm unsure about how you are computing when to update the voltages with the divider and FREQ variables, etc... What do you mean "minimum phase step each callback cycle" for `FREQ`? If I want a 10Hz sinusoid, can I set `FREQ = 10.0;` ?
7. I'm confused by the `drv/dac.c` function:
```C
void dac_set_voltage(uint8_t ch, double voltage, double max, double min)
```
Why does this accept a min and max? Is this always -10V and +10V? Can you document this somewhere?


Lastly, can you confirm that, after we merge this to `develop`, if someone checks out the latest code, they can simply follow the docs (i.e. use Tcl script to create Vivado project, build SDK code, progam board) and the AMDC works as it did before? Is the DAC related stuff disabled by default, and they must enable using `usr/user_config.h` system? Please make this the case.

BrandtBuchda committed:
Updated Documentation

BrandtBuchda committed:
Removed min max from dac_set_voltage

BrandtBuchda committed:
Added Status line direction macros

BrandtBuchda committed:
Updated Docs

BrandtBuchda committed:
Removed APP_DAC symbol to disable DAC Ex App

BrandtBuchda commented:
@npetersen2  @elsevers I've completed the changes as per Nathan's comments. I also removed the min and max parameter from the voltage and broadcast functions as they were unnecessary. I tested cloning a new repo from the branch with the my changes and it works as it did before. The DAC example app is only enabled when the `APP_DAC` symbol is added to the project symbols. I added directions for this in the example app documentation. For the driver to work the status lines and status ports have to be configured too. I added documentation for this in the  [driver docs](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/drv/docs/DAC-Expansion-Board.md).

npetersen2 reviewed (changes requested):
@BrandtBuchda much better documentation files, thanks for the updates!

- In [here](https://github.com/Severson-Group/AMDC-Firmware/blob/dac-revb-firmware/sdk/bare/common/drv/docs/DAC-Expansion-Board.md), the main title seems to be wrong, please fix.

I think pending this change, I am happy with this. There are a couple things I would like to see changed later (mainly to allow users to configure the mux things without having to update the `/common/` code), but for now, this is good. I will add issues so we remember to do these things.

BrandtBuchda committed:
Fixed main title

npetersen2 reviewed (approved):
Looks good to me. @elsevers Please review.

elsevers reviewed (approved):
Nice job @BrandtBuchda!
